### ISSUE TITLE
Data Loss: Empty directories not preserved through PUT/GET.

### ISSUE THREAD

**Issue by dan-da** [maidsafe/safe-api#440]
```
$ tree testdata/
testdata/
├── another.md
├── emptyfolder
├── noextension
├── subfolder
│   ├── sub2.md
│   └── subexists.md
└── test.md
```

```
$ safe files put --recursive testdata
FilesContainer created at: "safe://hnyynyiodw4extpc7xh3dncfgsg4sjzsygru9k8omo988brz688oxkxhxgbnc"
+  testdata/another.md              safe://hbhynydt5b95dmumcm8yig4u1keuuh8hgsr5yx39xn4mqikp91sbdhbpwp 
+  testdata/noextension             safe://hbyyyynw7fi7ko8u3wtai6eekko3m5y6u9ud7w1i4ajycmwfwu3wwpnn81 
+  testdata/subfolder/sub2.md       safe://hbhynydyr41tz98funce1eqzkgju8e1yb4343ppct4fiwie5n3qhsw7o7x 
+  testdata/subfolder/subexists.md  safe://hbhynyn9uodh1ju5uzyti3gmmtwburrssd89rcwcy3rzofdpypwomrzzte 
+  testdata/test.md                 safe://hbhynydpan7d94mwp1bun3mxfnrfrui131an7ihu11wsn8dkr8odab9qwn 
```

hmm, emptyfolder not mentioned in output of `safe files put --recursive`

```
$ safe files tree safe://hnyynyiodw4extpc7xh3dncfgsg4sjzsygru9k8omo988brz688oxkxhxgbnc
safe://hnyynyiodw4extpc7xh3dncfgsg4sjzsygru9k8omo988brz688oxkxhxgbnc
└── testdata
    ├── another.md
    ├── noextension
    ├── subfolder
    │   ├── sub2.md
    │   └── subexists.md
    └── test.md
```

emptyfolder not present when performing a GET either.   So we have lost the folder.  

I think that's a serious problem.

**Comment by dan-da** [comment]
Here is a another illustration/case, where we PUT just a single empty directory.  Note no trailing slash.

```
$ safe files put --recursive ../testdata/emptyfolder
FilesContainer created at: "safe://hnyynyii6kxymr5xd4z4k6qaag9a8pe9ghg4s7dn6167i3uirne847caa6bnc"
```

```
$ safe files ls safe://hnyynyii6kxymr5xd4z4k6qaag9a8pe9ghg4s7dn6167i3uirne847caa6bnc
Files of FilesContainer (version 0) at "safe://hnyynyii6kxymr5xd4z4k6qaag9a8pe9ghg4s7dn6167i3uirne847caa6bnc":
Files: 0   Size: 0   Total Files: 0   Total Size: 0
SIZE  CREATED  MODIFIED  NAME
```
```
$ safe cat safe://hnyynyii6kxymr5xd4z4k6qaag9a8pe9ghg4s7dn6167i3uirne847caa6bnc
Files of FilesContainer (version 0) at "safe://hnyynyii6kxymr5xd4z4k6qaag9a8pe9ghg4s7dn6167i3uirne847caa6bnc":
+------+------+---------+----------+------+
```

hmm, emptyfolder has vanished.

**Comment by dan-da** [comment]
Seems that empty dirs are presently unsupported.  See [commend in code](https://github.com/maidsafe/safe-api/blob/469f47f5703538716066723faf3c06bf8f170c41/safe-api/src/api/app/files.rs#L1104)

```
                        if metadata.is_dir() {
                            // Everything is in the iter. We dont need to recurse.
                            // so what do we do with dirs? decide if we want to support empty dirs also
                        }
```

So the question then becomes:  how to store empty dirs?

**Comment by theWebalyst** [comment]
Unless it is particularly difficult empty directors should be supported because that is expected behaviour and lots of things will rely on it.

Without this, every app that needs this has to do a lot of work to simulate it in the front end and it is a right pain in the bum!

**Comment by dan-da** [comment]
I plan to support empty directories, symlinks, file modes, and hidden files (.hidden) soon-ish.  high priority.  Metadata in FileDetails can help here.

**Comment by dan-da** [comment]
fixed by #542.  closing.
